File: leetcode/top_interview_150/two pointers/125. Valid Palindrome/main.go
package main

import (
	"fmt"
	"regexp"
	"slices"
	"strings"
)

func removeAllNonAlphanumeric(s string) string {
	reg := regexp.MustCompile(`[^a-zA-Z0-9]+`)

	clean := reg.ReplaceAllString(s, "")
	return clean
}

func isPalindrome(s string) bool {
	lower_string := strings.ToLower(s)
	prepare_string := removeAllNonAlphanumeric(lower_string)

	forward_slice := strings.Split(prepare_string, " ")

	backward_slice := strings.Split(prepare_string, " ")
	slices.Reverse(backward_slice)

	fmt.Println(forward_slice, backward_slice)

	forward := strings.Join(forward_slice, "")
	backward := strings.Join(backward_slice, "")

	fmt.Println(forward, backward)

	return forward == backward
}

var ()

func main() {
	fmt.Println(isPalindrome("race a car"))
}

File: leetcode/top_interview_150/two pointers/125. Valid Palindrome/main.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s) - 1
        while left < right:
            print(left, right)
            if not s[left].isalnum():
                left += 1
                continue
            if not s[right].isalnum():
                right -= 1
                continue
            if s[left].lower() == s[right].lower():
                left += 1
                right -= 1
                continue
            else:
                return False
        return True


if __name__ == "__main__":
    print(Solution().isPalindrome("race a car"))
